---
title: "sangeranalyseR"
author: "Author: Kuan-Hao Chao (u7022787@anu.edu.au)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output: 
  BiocStyle::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_caption: yes
#bibliography: bibliography.bib
fontsize: 14pt
vignette: >
  %\VignetteIndexEntry{sangeranalyseR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignettePackage{sangeranalyseR}
---
```{r style, echo=FALSE, results="asis", message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

# Introduction

## Design Overview
The [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) provides S4 classes for importing Sanger sequencing files into R environment. 

### S4 Classes
There are three main S4 classes in [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) which are **'SangerSingleRead'**, **'SangerConsensusRead'** and **'AlignedConsensusSet'**. Each of them stores different level of input parameters and results in Sanger sequencing data analysis.

* **'SangerSingleRead'** extends from 'sangerseq' class and stores raw data 'abif' class in [sangerseqR](https://bioconductor.org/packages/release/bioc/html/sangerseqR.html) as well as essential information including quality trimming and chromatogram parameters. It corresponds to a single ABIF file in Sanger sequencing.

* **'SangerConsensusRead'** contains two lists of 'SangerSingleRead' which are forward and reverse reads and their alignment results. It corresponds to a contig in Sanger sequencing.

* **'AlignedConsensusSet'** contains a list of 'SangerConsensusRead' and the alignment results for all consensus reads. It corresponds to the rebuild fragment of genome in Sanger sequencing.

### Interactive Shiny UI
We build two local interactive Shiny user interface for users to explore their created **'SangerConsensusRead'** and **'AlignedConsensusSet'**. Inside the Shiny UI, users can change the input parameters for quality trimming, chromatogram and rebuild consensus read etc. Information showed on th UI will be dynamic based on user input. Furthermore, all changed parameters and results will be applied to a new S4 instance and be exported as an *RDS* file.

## System Requirement

## Installation
### Installing from GitHub using devtools
```{r eval=FALSE}
install.packages("devtools")
library(devtools)
install_github("roblanf/sangeranalyseR")
```

### Installing from Bioconductor




# S4 Classes
## SangerSingleRead
### ABIF Filename Regulation
### S4 Instance Creation
First, we load [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) in to R environment.
```{r}
library(sangeranalyseR)
```

Then, we find the absolute filename for target ABIF file.
```{r}
inputFilesPath <- system.file("extdata/", package = "sangeranalyseR")

A_chloroticaFdReadFN <- file.path(inputFilesPath,                                                 
                                  "Allolobophora_chlorotica",
                                  "RBNII396-13[C_LepFolF,C_LepFolR]_F_1.ab1")
```

Last, we create **'SangerSingleRead'** S4 instance by running new function.
```{r}
A_chloroticaSingleRead <- new("SangerSingleRead",
                               readFeature           = "Forward Read",
                               readFileName          = A_chloroticaFdReadFN,
                               TrimmingMethod        = "M2",
                               M1TrimmingCutoff      = NULL,
                               M2CutoffQualityScore  = 40,
                               M2SlidingWindowSize   = 10,
                               baseNumPerRow         = 100,
                               heightPerRow          = 200,
                               signalRatioCutoff     = 0.33,
                               showTrimmed           = TRUE)
```

Furthermore, we have created a constructor wrapper function for **'SangerSingleRead'**. The result of new and constructor creation methods are same.
```{r}
 A_chloroticaSingleRead <- SangerSingleRead(readFeature           = "Forward Read",
                                            readFileName          = A_chloroticaFdReadFN,
                                            geneticCode           = GENETIC_CODE,
                                            TrimmingMethod        = "M2",
                                            M1TrimmingCutoff      = NULL,
                                            M2CutoffQualityScore  = 40,
                                            M2SlidingWindowSize   = 10,
                                            baseNumPerRow         = 100,
                                            heightPerRow          = 200,
                                            signalRatioCutoff     = 0.33,
                                            showTrimmed           = TRUE)
```

### Inputs Definition
The inputs of new and constructor methods are same. Both their input parameters can be divied into three main categories:

#### Basic Parameters
* `readFeature` : Specify whether the read is *forward* or *reverse* reads.
* `readFileName`          = A_chloroticaFdReadFN
* `geneticCode`           = GENETIC_CODE

#### Quality-trimming Related Parameters
* `TrimmingMethod`        = "M2"
* `M1TrimmingCutoff`      = NULL
* `M2CutoffQualityScore`  = 40
* `M2SlidingWindowSize`   = 10

#### Chromatogram Related Parameters
* `baseNumPerRow`         = 100
* `heightPerRow`          = 200
* `signalRatioCutoff`     = 0.33
* `showTrimmed`           = TRUE



### Slots Definition

After creating **'SangerSingleRead'**, we can 
A_chloroticaSingleRead
```{r eval=FALSE}
A_chloroticaSingleRead@readFeature
A_chloroticaSingleRead@readFileName
A_chloroticaSingleRead@abifRawData
A_chloroticaSingleRead@QualityReport
A_chloroticaSingleRead@ChromatogramParam
A_chloroticaSingleRead@primaryAASeq
A_chloroticaSingleRead@geneticCode
A_chloroticaSingleRead@primarySeq
A_chloroticaSingleRead@secondarySeq
A_chloroticaSingleRead@traceMatrix
A_chloroticaSingleRead@peakPosMatrix
A_chloroticaSingleRead@peakAmpMatrix
```
#### QualityReport
#### ChromatogramParam


## SangerConsensusRead
### Inputs Definition
### Slots Definition
### Example

## AlignedConsensusSet
### Inputs Definition
### Slots Definition
### Example

# Shiny App
## ShinySangerConsensusRead
### Example
* UI for **'SangerConsensusRead'** class: Users can access to each read and an overview page for consensus reads. 

launchAppConsensusRead 

## ShinySangerAlignedConsensusSet
### Example
* UI for **'AlignedConsensusSet'** class: Users can recursively access to all single reads, each overview page for consensus reads and 
launchAppAlignedConsensusSet


# Conclusion

# Session Information

# Reference
