---
title: "sangeranalyseR"
author: "Author: Kuan-Hao Chao (u7022787@anu.edu.au)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output: 
  BiocStyle::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_caption: yes
#bibliography: bibliography.bib
fontsize: 14pt
vignette: >
  %\VignetteIndexEntry{sangeranalyseR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignettePackage{sangeranalyseR}
---
```{r style, echo=FALSE, results="asis", message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

# Introduction

## Design Overview
The [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) provides S4 classes for importing Sanger sequencing files into R environment. 

### S4 Classes
There are three main S4 classes in [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) which are **'SangerSingleRead'**, **'SangerConsensusRead'** and **'AlignedConsensusSet'**. Each of them stores different level of input parameters and results in Sanger sequencing data analysis.

* **'SangerSingleRead'** extends from 'sangerseq' class and stores raw data 'abif' class in [sangerseqR](https://bioconductor.org/packages/release/bioc/html/sangerseqR.html) as well as essential information including quality trimming and chromatogram parameters. It corresponds to a single ABIF file in Sanger sequencing.

* **'SangerConsensusRead'** contains two lists of 'SangerSingleRead' which are forward and reverse reads and their alignment results. It corresponds to a contig in Sanger sequencing.

* **'AlignedConsensusSet'** contains a list of 'SangerConsensusRead' and the alignment results for all consensus reads. It corresponds to the rebuild fragment of genome in Sanger sequencing.

### Interactive Shiny UI
We build two local interactive Shiny user interface for users to explore their created **'SangerConsensusRead'** and **'AlignedConsensusSet'**. Inside the Shiny UI, users can change the input parameters for quality trimming, chromatogram and rebuild consensus read etc. Information showed on th UI will be dynamic based on user input. Furthermore, all changed parameters and results will be applied to a new S4 instance and be exported as an *RDS* file.

## System Requirement

## Installation
### Installing from GitHub using devtools
```{r eval=FALSE}
install.packages("devtools")
library(devtools)
install_github("roblanf/sangeranalyseR")
```

### Installing from Bioconductor
>>> After uploading to Bioconductor







# S4 Classes
## SangerSingleRead

### ABIF Filename Regulation
The figure below shows the filename regulation of input raw ABIF file which is going to be used as input to create **"SangerSingleRead"**. All raw ABIF files must be *forward* or *reverse* and the reason there is a [*Number Index*] in the suffix is that later we are going to create **"SangerConsensusRead"** S4 instance. All reads that are going to be aligned in the same consensus group must be indexed properly.

```{r fig.width=10, fig.height=1, echo=FALSE}
library(png)
library(grid)
img <- readPNG("figure/SangerSingleRead_filename_regulation.png")
grid.raster(img, just = "center")
```

For example, "RBNII396-13[C_LepFolF,C_LepFolR]_F_1.ab1" is a valid name. 

* The **[*Consensus Read Name*]** is *"RBNII396-13[C_LepFolF,C_LepFolR]"*
* The **[*F,R*]** is *"F"* 
* The **[*Number Index*]** is *"1"*.

### S4 Instance Creation Example
First, we load [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) into R environment.
```{r}
library(sangeranalyseR)
```

Then, we find the absolute filename for target ABIF file.
```{r}
inputFilesPath <- system.file("extdata/", package = "sangeranalyseR")

A_chloroticaFdReadFN <- file.path(inputFilesPath,                                                 
                                  "Allolobophora_chlorotica",
                                  "RBNII396-13[C_LepFolF,C_LepFolR]_F_1.ab1")
```

Last but not least, we create **'SangerSingleRead'** S4 instance by running *"new"* function.
```{r}
A_chloroticaSingleRead <- new("SangerSingleRead",
                               readFeature           = "Forward Read",
                               readFileName          = A_chloroticaFdReadFN,
                               TrimmingMethod        = "M2",
                               M1TrimmingCutoff      = NULL,
                               M2CutoffQualityScore  = 40,
                               M2SlidingWindowSize   = 10,
                               baseNumPerRow         = 100,
                               heightPerRow          = 200,
                               signalRatioCutoff     = 0.33,
                               showTrimmed           = TRUE)
```

Furthermore, we also create a constructor wrapper function for **'SangerSingleRead'**. The created instance of new and constructor creation methods are same.
```{r}
 A_chloroticaSingleRead <- SangerSingleRead(readFeature           = "Forward Read",
                                            readFileName          = A_chloroticaFdReadFN,
                                            geneticCode           = GENETIC_CODE,
                                            TrimmingMethod        = "M2",
                                            M1TrimmingCutoff      = NULL,
                                            M2CutoffQualityScore  = 40,
                                            M2SlidingWindowSize   = 10,
                                            baseNumPerRow         = 100,
                                            heightPerRow          = 200,
                                            signalRatioCutoff     = 0.33,
                                            showTrimmed           = TRUE)
```

### Inputs Definition
The inputs of new and constructor methods are same. Both of their input parameters can be divied into three main categories:

#### Basic Parameters
* **`readFeature`** : Specify whether the read is forward or reverse reads. The value must be *"Forward Read"* or *"Reverse Read"*.
* **`readFileName`** : The absolute filename of the ABIF file.
* **`geneticCode`** : Named character vector in the same format as GENETIC_CODE (the default), which represents the standard genetic code. This is the code with which the function will attempt to translate your DNA sequences. Users can get an appropriate vector with the getGeneticCode() function. The default is the standard code.

#### Quality-trimming Related Parameters
There are two read trimming methods for users to choose from. The Method 1 (M1) is *"Logarithmic Scale Trimming"* and the Method 2 (M2) is *"Logarithmic Scale Sliding Window Trimming"*. 

Method 1 (M1) is (!!!!!!! Add !!!!!!!)

Method 2 (M2) is similar to "Trimmomatic"(Bolger et al.) SLIDINGWINDOW trimming method. We take the average logarithmic quality score in the sliding window size and if the average quality score is smaller than the cutoff quality score, then the first base pair of the sliding window will be the trimming start point.

* **`TrimmingMethod`** : Specify the read trimming method for this SangerSingleRead. The value must be *"M1"* (the default) or *'M2'*.
* **`M1TrimmingCutoff`** : The trimming cutoff for the Method 1. If `TrimmingMethod` is *"M1"*, then the default value is *"0.0001"*. Otherwise, the value must be *"NULL"*
* **`M2CutoffQualityScore`** : The trimming cutoff quality score for the Method 2. If `TrimmingMethod` is *"M2"*, then the default value is *"20"*. Otherwise, the value must be *"NULL"*. It works with `M2SlidingWindowSize`.
* **`M2SlidingWindowSize`** : The trimming sliding window size for the Method 2. If `TrimmingMethod` is *"M2"*, then the default value is *"5"*. Otherwise, the value must be *"NULL"*. It works with `M2CutoffQualityScore`.

#### Chromatogram Related Parameters

These parameters are used for drawing chromatogram in Shiny UI.

* **`baseNumPerRow`** : It defines maximum base pairs in each row. The default value is *"100"*
* **`heightPerRow`** : It defines the height of each row in chromatogram. The default value is *"200"*.
* **`signalRatioCutoff`** : The ratio of the height of a secondary peak to a primary peak. Secondary peaks higher than this ratio are annotated. Those below the ratio are excluded. The default value is *"0.33"*.
* **`showTrimmed`** : The logical value storing whether to show trimmed base pairs in chromatogram. The default value is *"TRUE"*.

### Slots Definition

**'SangerSingleRead'** inherits 'sangerseq' class. It contains all slots in 'sangerseq', so every function that uses 'sangerseq' as input can be replaced by **'SangerSingleRead'**. For example, we can use **'SangerSingleRead'** to run *"chromatogram"*, *'makeBaseCalls'* functions in [sangerseqR](https://bioconductor.org/packages/release/bioc/html/sangerseqR.html).
For inherited slots in **'SangerSingleRead'**, please refer to [sangerseqR reference manual](https://bioconductor.org/packages/release/bioc/manuals/sangerseqR/man/sangerseqR.pdf). Following are other slots in **'SangerSingleRead'**.

* **`readFeature`** : *"Forward Read"* or *"Reverse Read"*
* **`readFileName`** : Absolute filename of ABIF file.
* **`abifRawData`** : 'abif' class in [sangerseqR](https://bioconductor.org/packages/release/bioc/html/sangerseqR.html)
* **`QualityReport`** : S4 class defined in this package storing read quality-related parameters.
* **`ChromatogramParam`** : S4 class defined in this package storing read chromatogram-related parameters.
* **`geneticCode`** : Named character vector in the same format as GENETIC_CODE (the default), which represents the standard genetic code.

Here, we show all slots in **'SangerSingleRead'**
```{r }
slotNames(A_chloroticaSingleRead)
```

## SangerConsensusRead
### ABIF Filename Regulation & Input directory regulation
### S4 Instance Creation Example

```{r}
inputFilesPath <- system.file("extdata/", package = "sangeranalyseR")

A_chloroticaFdReadFN <- file.path(inputFilesPath,                                                 
                                  "Allolobophora_chlorotica",
                                  "RBNII396-13[C_LepFolF,C_LepFolR]_F_1.ab1")
```

### Inputs Definition
### Slots Definition

## AlignedConsensusSet
### ABIF Filename Regulation & Input directory regulation
### S4 Instance Creation Example
### Inputs Definition
### Slots Definition

# Shiny App
## ShinySangerConsensusRead
### Example
* UI for **'SangerConsensusRead'** class: Users can access to each read and an overview page for consensus reads. 

launchAppConsensusRead 

## ShinySangerAlignedConsensusSet
### Example
* UI for **'AlignedConsensusSet'** class: Users can recursively access to all single reads, each overview page for consensus reads and 
launchAppAlignedConsensusSet


# Conclusion

# Session Information

# Reference
Bolger, A. M., Lohse, M., & Usadel, B. (2014). Trimmomatic: a flexible trimmer for Illumina sequence data. Bioinformatics, 30(15), 2114-2120.
