---
title: "An Introduction to sangeranalyseR"
author: "Author: Kuan-Hao Chao (u7022787@anu.edu.au)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output: 
  BiocStyle::html_document:
    toc_float: true
    code_folding: show
    fig_caption: yes
  BiocStyle::pdf_document: default
#bibliography: bibliography.bib
fontsize: 14pt
vignette: >
  %\VignetteIndexEntry{sangeranalyseR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
  %\VignettePackage{sangeranalyseR}
---
```{r style, echo=FALSE, results="asis", message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE)
```

```{r}
library(sangeranalyseR)
```

# Introduction
[sangeranalyseR](https://github.com/roblanf/sangeranalyseR) is an R package for analysing Sanger sequencing reads, especially those from ABIF platform, in pure R environment. 
There are three levels in [sangeranalyseR](https://github.com/roblanf/sangeranalyseR) which are *'SangerRead'*, *'SangerContig'* and *'SangerAlignment'*. Users can choose which level to start the analysis.

In this documentation, we intoduce analysis steps in three levels and main features with examples.


# SangerRead
**'SangerRead'** extends from 'sangerseq' class and stores raw data 'abif' class in [sangerseqR](https://bioconductor.org/packages/release/bioc/html/sangerseqR.html) as well as essential information including quality trimming and chromatogram parameters. It corresponds to a single ABIF file in Sanger sequencing. 

## Create Sanger Object 

```{r}
inputFilesPath <- system.file("extdata/", package = "sangeranalyseR")

A_chloroticaFdReadFN <- file.path(inputFilesPath,                                                 
                                  "Allolobophora_chlorotica",
                                  "RBNII396-13[C_LepFolF,C_LepFolR]_F_1.ab1")
```
```{r}
singleRead <- SangerRead(readFeature           = "Forward Read",
                         readFileName          = A_chloroticaFdReadFN)
```

## Write FASTA file
```{r}
writeFastaSR(singleRead)
```

## Generate Report
```{r eval=FALSE}
generateReportSR(singleRead)
```

# SangerContig
**'SangerContig'** contains two lists of 'SangerRead' which are forward and reverse read list as well as the alignment results. It corresponds to a contig in Sanger sequencing.

## Create Sanger Object 
```{r}
rawDataDir <- system.file("extdata", package = "sangeranalyseR")
inputFilesParentDir <- file.path(rawDataDir, "Allolobophora_chlorotica")
contigName <- "ACHLO006-09[LCO1490_t1,HCO2198_t1]"
suffixForwardRegExp <- "_[F]_[0-9]*.ab1"
suffixReverseRegExp <- "_[R]_[0-9]*.ab1"
```
```{r}
sangerContig <- SangerContig(parentDirectory      = inputFilesParentDir,
                             contigName           = contigName,
                             suffixForwardRegExp  = suffixForwardRegExp,
                             suffixReverseRegExp  = suffixReverseRegExp)
```

## Launch Shiny App 
```{r eval=FALSE}
launchAppSC(sangerContig)
```

## Write FASTA file
```{r}
writeFastaSC(sangerContig)
```

## Generate Report
```{r eval=FALSE}
generateReportSC(sangerContig)
```

# SangerAlignment
**'SangerAlignment'** contains a list of 'SangerContig' and the alignment results for all contigs. It corresponds to a rebuild DNA sequence fragment in Sanger sequencing. 

## Create Sanger Object 
```{r}
inputFilesParentDir <- system.file("extdata", package = "sangeranalyseR")
suffixForwardRegExp <- "_[F]_[0-9]*.ab1"
suffixReverseRegExp <- "_[R]_[0-9]*.ab1"
```

```{r  eval=FALSE}
sangerAlignment <- SangerAlignment(parentDirectory = inputFilesParentDir,
                                         suffixForwardRegExp = suffixForwardRegExp,
                                         suffixReverseRegExp = suffixReverseRegExp,)
```

## Launch Shiny App 
```{r eval=FALSE}
launchAppSA(sangerAlignment)
```

## Write FASTA file
```{r  eval=FALSE}
writeFastaSA(sangerAlignment)
```

## Generate Report
```{r eval=FALSE}
generateReportSA(sangerAlignment)
```
