

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Advanced User Guide - SangerAlignment (FASTA) &mdash; sangeranalyseR  documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="sangeranalyseR  documentation" href="../index.html"/>
        <link rel="next" title="Conclusion" href="conclusion.html"/>
        <link rel="prev" title="Advanced User Guide - SangerContig (FASTA)" href="advance_user_SangerContig_fasta.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> sangeranalyseR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginners Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_ab1.html">Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_ab1.html">Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_ab1.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_fasta.html">Advanced User Guide - <em>SangerRead</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_fasta.html">Advanced User Guide - <em>SangerContig</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-sangeralignment-fasta-input">Preparing <em>SangerAlignment</em> <strong>FASTA</strong> input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sangeralignment-instance-from-fasta">Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#writing-sangeralignment-fasta-files-fasta">Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-sangeralignment-report-fasta">Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="contact.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">sangeranalyseR</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/content/advance_user_SangerAlignment_fasta.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="advanced-user-guide-sangeralignment-fasta">
<h1>Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)<a class="headerlink" href="#advanced-user-guide-sangeralignment-fasta" title="Permalink to this headline">¶</a></h1>
<p><em>SangerAlignment</em> is the highest class level in sangeranalyseR showed in <a class="reference internal" href="#sangeralignment-hierachy-fasta"><span class="std std-ref">Figure_1</span></a>. It contains <em>SangerContig</em> list and the contigs alignment result. Users can access to <em>SangerContig</em> and <em>SangerRead</em> instance inside <em>SangerAlignment</em> instance. In this section, we are going to go through detailed sangeranalyseR data analysis steps in <em>SangerAlignment</em> level from <strong>FASTA</strong> file input.</p>
<div class="figure align-center" id="id1">
<span id="sangeralignment-hierachy-fasta"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_hierachy.png"><img alt="../_images/SangerAlignment_hierachy.png" src="../_images/SangerAlignment_hierachy.png" style="width: 545.4px; height: 218.8px;" /></a>
<p class="caption"><span class="caption-text">Figure 1. Classes hierarchy in sangeranalyseR, <em>SangerAlignment</em> level.</span></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="preparing-sangeralignment-fasta-input">
<h2>Preparing <em>SangerAlignment</em> <strong>FASTA</strong> input<a class="headerlink" href="#preparing-sangeralignment-fasta-input" title="Permalink to this headline">¶</a></h2>
<p>We design the <strong>FASTA</strong> file input for those who do not want to do quality trimming and base calling for each <em>SangerRead</em> in <em>SangerAlignment</em>; therefore, it does not contain quality trimming and chromatogram input parameters and results in <em>SangerRead</em> slots.</p>
<p>Before starting the analysis, users need to prepare one <strong>FASTA</strong> file containing sequence of all reads. Inside the <strong>FASTA</strong> file, there are strings starting with &#8220;&gt;&#8221; before each read which are the read names. Because sangeranalyseR will group reads into &#8220;Forward Read List&#8221; and &#8220;Reverse Read List&#8221;, users have to follow the naming regulations for the read names. Below are some regulations:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Reads that are in the same contig have to share the same contig name.</li>
<li>Forward or reverse direction also has to be specified in the read names.</li>
</ul>
</div>
<p>There are three parameters, <code class="code docutils literal"><span class="pre">fastaFileName</span></code>, <code class="code docutils literal"><span class="pre">suffixForwardRegExp</span></code>, and <code class="code docutils literal"><span class="pre">suffixReverseRegExp</span></code>, that users need to provide so that program can automatically group all <strong>FASTA</strong> files.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li><code class="code docutils literal"><span class="pre">fastaFileName</span></code>: The <strong>FASTA</strong> file that contains sequence of all reads. The read names have to follow the naming regulation.</li>
<li><code class="code docutils literal"><span class="pre">suffixForwardRegExp</span></code>: The value of this parameter is a regular expression that matches all filenames in forward direction. <code class="code docutils literal"><span class="pre">grepl</span></code> function in R is used to select forward reads from all <strong>FASTA</strong> files.</li>
<li><code class="code docutils literal"><span class="pre">suffixReverseRegExp</span></code>: The value of this parameter is a regular expression that matches all filenames in reverse direction. <code class="code docutils literal"><span class="pre">grepl</span></code> function in R is used to select reverse reads from all <strong>FASTA</strong> files.</li>
</ul>
</div>
<p>No doubt read names in the original <strong>FASTA</strong> file will not follow the naming regulation; however, it is highly not recommended to change the name directly in the raw <strong>FASTA</strong> file. Therefore, we provide a feature to let users do read names mapping conversion by a <strong>CSV</strong> file showed in <a class="reference internal" href="#sangeralignment-read-names-conversion"><span class="std std-ref">Figure_2</span></a>. The first column is &#8220;original_read_name&#8221; which are the read names in the raw <strong>FASTA</strong> file, and the second column is &#8220;analysis_read_name&#8221; which are the read names that follow the naming regulation. The read names will be mapped onto the names in &#8220;original_read_name&#8221; without changing the raw <strong>FASTA</strong> file. <code class="code docutils literal"><span class="pre">namesConversionCSV</span></code> is the parameter that stores the path to this <strong>CSV</strong> file.</p>
<div class="figure align-center" id="id2">
<span id="sangeralignment-read-names-conversion"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_read_names_conversion.png"><img alt="../_images/SangerAlignment_read_names_conversion.png" src="../_images/SangerAlignment_read_names_conversion.png" style="width: 725.6px; height: 239.2px;" /></a>
<p class="caption"><span class="caption-text">Figure 2. <em>SangerAlignment</em> <strong>CSV</strong> file - read names conversion.</span></p>
</div>
<p>Here, we have another more complicated example.</p>
<div class="figure align-center" id="id3">
<span id="sangeralignment-fasta-input"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_fasta_input.png"><img alt="../_images/SangerAlignment_fasta_input.png" src="../_images/SangerAlignment_fasta_input.png" style="width: 852.0px; height: 472.8px;" /></a>
<p class="caption"><span class="caption-text">Figure 3. <em>SangerAlignment</em> <strong>FASTA</strong> input file.</span></p>
</div>
<p><a class="reference internal" href="#sangeralignment-fasta-input"><span class="std std-ref">Figure_3</span></a> shows the <strong>FASTA</strong> input file and the read names in it will be mapping conversed by <strong>CSV</strong> file showed in <a class="reference internal" href="#sangeralignment-read-names-conversion"><span class="std std-ref">Figure_2</span></a> (Only two reads are showed). sangeranalyseR will first match the <code class="code docutils literal"><span class="pre">contigName</span></code> to exclude unrelated reads. The direction of reads in each contig will be grouped by matching <code class="code docutils literal"><span class="pre">suffixForwardRegExp</span></code> and <code class="code docutils literal"><span class="pre">suffixReverseRegExp</span></code> with read names. Therefore, it is important to carefully select <code class="code docutils literal"><span class="pre">suffixForwardRegExp</span></code> and <code class="code docutils literal"><span class="pre">suffixReverseRegExp</span></code>. The bad file naming and wrong regex matching might accidentally include reverse reads into the forward read list or vice versa, which will make the program generate totally wrong results. Therefore, users should have a consistent naming strategy. In this example, &#8220;<code class="code docutils literal"><span class="pre">_[0-9]+_F</span></code>&#8221;, &#8220;<code class="code docutils literal"><span class="pre">_[0-9]+_R</span></code>&#8221; for matching forward and reverse reads are highly suggested and are used as default. It is a good habit to index your reads in the same contig group because there might be more than one read that are in the forward or reverse direction.</p>
<div class="figure align-center" id="id4">
<span id="sangeranalyser-filename-convention-sangeralignment-fasta"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_filename_convention_fasta.png"><img alt="../_images/sangeranalyseR_filename_convention_fasta.png" src="../_images/sangeranalyseR_filename_convention_fasta.png" style="width: 538.25px; height: 28.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 4. Suggested read naming regulation in <strong>FASTA</strong> file - <em>SangerAlignment</em>.</span></p>
</div>
<p><a class="reference internal" href="#sangeranalyser-filename-convention-sangeralignment-fasta"><span class="std std-ref">Figure_4</span></a> shows the suggested reads naming regulation. Users are strongly recommended to follow this reads naming regulation and use the default <code class="code docutils literal"><span class="pre">suffixForwardRegExp</span></code> : &#8220;<code class="code docutils literal"><span class="pre">_[0-9]+_F</span></code>&#8221; and <code class="code docutils literal"><span class="pre">suffixReverseRegExp</span></code> : &#8220;<code class="code docutils literal"><span class="pre">_[0-9]+_R</span></code>&#8221; to reduce any chance of error.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="creating-sangeralignment-instance-from-fasta">
<h2>Creating <em>SangerAlignment</em> instance from <strong>FASTA</strong><a class="headerlink" href="#creating-sangeralignment-instance-from-fasta" title="Permalink to this headline">¶</a></h2>
<p>After preparing the input directory, we can create the <em>SangerAlignment</em> S4 instance by running <code class="code docutils literal"><span class="pre">SangerAlignment</span></code> constructor function or <code class="code docutils literal"><span class="pre">new</span></code> method. The constructor function is a wrapper for <code class="code docutils literal"><span class="pre">new</span></code> method and it makes instance creation more intuitive. Most parameters in the constructor have their own default values. In the constructor below, we list important parameters.</p>
<div class="highlight-R"><div class="highlight"><pre><span></span>sangerAlignmentFa <span class="o">&lt;-</span> SangerAlignment<span class="p">(</span>inputSource           <span class="o">=</span> <span class="s">&quot;FASTA&quot;</span><span class="p">,</span>
                                     fastaFileName         <span class="o">=</span> <span class="s">&quot;Sanger_all_reads.fa&quot;</span><span class="p">,</span>
                                     namesConversionCSV    <span class="o">=</span> <span class="s">&quot;names_conversion.csv&quot;</span><span class="p">,</span>
                                     suffixForwardRegExp   <span class="o">=</span> <span class="s">&quot;_[0-9]+_F&quot;</span><span class="p">,</span>
                                     suffixReverseRegExp   <span class="o">=</span> <span class="s">&quot;_[0-9]+_R&quot;</span><span class="p">,</span>
                                     refAminoAcidSeq       <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span>
                                     <span class="p">)</span>
</pre></div>
</div>
<p>The inputs of <code class="code docutils literal"><span class="pre">SangerAlignment</span></code> constructor function and <code class="code docutils literal"><span class="pre">new</span></code> method are same. For more details about <em>SangerAlignment</em> inputs and slots definition, please refer to <a class="reference external" href="http://packages.python.org/an_example_pypi_project/">sangeranalyseR reference manual (need update)</a>.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="writing-sangeralignment-fasta-files-fasta">
<h2>Writing <em>SangerAlignment</em> FASTA files <sub>(FASTA)</sub><a class="headerlink" href="#writing-sangeralignment-fasta-files-fasta" title="Permalink to this headline">¶</a></h2>
<p>Users can write the <em>SangerAlignment</em> instance to <strong>FASTA</strong> files. There are four options for users to choose from in <code class="code docutils literal"><span class="pre">selection</span></code> parameter.</p>
<ul class="simple">
<li><code class="code docutils literal"><span class="pre">contigs_unalignment</span></code>: Writing contigs into a single <strong>FASTA</strong> file.</li>
<li><code class="code docutils literal"><span class="pre">contigs_alignment</span></code>: Writing contigs alignment and contigs consensus read to a single <strong>FASTA</strong> file.</li>
<li><code class="code docutils literal"><span class="pre">all_reads</span></code>: Writing all reads to a single <strong>FASTA</strong> file.</li>
<li><code class="code docutils literal"><span class="pre">all</span></code>: Writing contigs, contigs alignment, and all reads into three different files.</li>
</ul>
<p>Below is the one-line function that users need to run. This function mainly depends on <code class="code docutils literal"><span class="pre">writeXStringSet</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> R package. Users can set the compression level through <code class="code docutils literal"><span class="pre">writeFasta</span></code> function.</p>
<div class="highlight-R"><div class="highlight"><pre><span></span>writeFasta<span class="p">(</span>sangerAlignmentFa<span class="p">,</span>
           outputDir         <span class="o">=</span> <span class="kp">tempdir</span><span class="p">(),</span>
           compress          <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
           compression_level <span class="o">=</span> <span class="kc">NA</span><span class="p">,</span>
           selection         <span class="o">=</span> <span class="s">&quot;all&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can download the output FASTA file of this example through the following three links:</p>
<ul class="simple">
<li><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/Sanger_contigs_unalignment.fa">contigs_unalignment FASTA file</a></li>
<li><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/Sanger_contigs_alignment.fa">contigs_alignment FASTA file</a></li>
<li><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/Sanger_all_trimmed_reads.fa">all_reads FASTA file</a></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="generating-sangeralignment-report-fasta">
<h2>Generating <em>SangerAlignment</em> report <sub>(FASTA)</sub><a class="headerlink" href="#generating-sangeralignment-report-fasta" title="Permalink to this headline">¶</a></h2>
<p>Last but not least, users can save <em>SangerAlignment</em> instance into a report after the analysis. The report will be generated in <strong>HTML</strong> by knitting <strong>Rmd</strong> files. There are two parameters, <code class="code docutils literal"><span class="pre">includeSangerContig</span></code> and <code class="code docutils literal"><span class="pre">includeSangerRead</span></code>, for users to decide to which level the <em>SangerAlignment</em> report will go. Moreover, after the reports are generated, users can easily navigate through reports in different levels within the <strong>HTML</strong> file.</p>
<ul class="simple">
<li><code class="code docutils literal"><span class="pre">includeSangerContig</span></code>: Whether users want to generate the report of each <em>SangerContig</em> in <em>SangerAlignment</em>.</li>
<li><code class="code docutils literal"><span class="pre">includeSangerRead</span></code>: If <code class="code docutils literal"><span class="pre">includeSangerContig</span></code> is <code class="code docutils literal"><span class="pre">TRUE</span></code>, then users can set this value to decide whether they want to include <em>SangerRead</em> reports in each <em>SangerContig</em>.</li>
</ul>
<p>One thing to pay attention to is that if users have many reads, it will take quite a long time to write out all reports. If users only want to generate the <em>SangerAlignment</em> level (contigs alignment), remember to set <code class="code docutils literal"><span class="pre">includeSangerContig</span></code> and <code class="code docutils literal"><span class="pre">includeSangerRead</span></code> to <code class="code docutils literal"><span class="pre">FALSE</span></code> in order to save time.</p>
<div class="highlight-R"><div class="highlight"><pre><span></span>generateReport<span class="p">(</span>sangerAlignmentFa<span class="p">,</span>
               outputDir           <span class="o">=</span> <span class="kp">tempdir</span><span class="p">(),</span>
               includeSangerContig <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
               includeSangerRead   <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can access to &#8216;<em>Basic Information</em>&#8216;, &#8216;<em>Contigs Consensus</em>&#8216;, &#8216;<em>Contigs Alignment</em>&#8216; and &#8216;<em>Contigs Tree</em>&#8216; sections inside the generated <a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/FASTA/SangerAlignment/SangerAlignment_Report.html">SangerContig html report of this example</a>. Furthermore, users can also navigate through html reports of all forward and reverse <em>SangerRead</em> in this <em>SangerContig</em> report.</p>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="conclusion.html" class="btn btn-neutral float-right" title="Conclusion" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="advance_user_SangerContig_fasta.html" class="btn btn-neutral" title="Advanced User Guide - SangerContig (FASTA)" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Kuan-Hao Chao.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>