

<!DOCTYPE html>
<html class="writer-html5" lang="English" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Advanced User Guide - SangerRead (AB1) &mdash; sangeranalyseR  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced User Guide - SangerContig (AB1)" href="advance_user_SangerContig_ab1.html" />
    <link rel="prev" title="Beginners Guide" href="beginner.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> sangeranalyseR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginners Guide</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-sangerread-ab1-input">Preparing <em>SangerRead</em> <strong>AB1</strong> input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sangerread-instance-from-ab1">Creating <em>SangerRead</em> instance from <strong>AB1</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualizing-sangerread-trimmed-read">Visualizing <em>SangerRead</em> trimmed read</a></li>
<li class="toctree-l2"><a class="reference internal" href="#updating-sangerread-quality-trimming-parameters">Updating <em>SangerRead</em> quality trimming parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#writing-sangerread-fasta-file-ab1">Writing <em>SangerRead</em> FASTA file <sub>(AB1)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-sangerread-report-ab1">Generating <em>SangerRead</em> report <sub>(AB1)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#a-reproducible-example-sangerread-ab1">A Reproducible Example (<em>SangerRead</em>, <strong>ab1</strong>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">1. Preparing <em>SangerRead</em> <strong>AB1</strong> input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">2. Creating <em>SangerRead</em> instance from <strong>AB1</strong></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">3. Visualizing <em>SangerRead</em> trimmed read</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">4. Writing <em>SangerRead</em> FASTA file <sub>(AB1)</sub></a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">5. Generating <em>SangerRead</em> report <sub>(AB1)</sub></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_ab1.html">Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_ab1.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_fasta.html">Advanced User Guide - <em>SangerRead</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_fasta.html">Advanced User Guide - <em>SangerContig</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_fasta.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_page.html">How to ...</a></li>
<li class="toctree-l1"><a class="reference internal" href="function_manual.html">User Manual (functions)</a></li>
<li class="toctree-l1"><a class="reference internal" href="frequent_questions.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="contact.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sangeranalyseR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/content/advance_user_SangerRead_ab1.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="advanced-user-guide-sangerread-ab1">
<h1>Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)<a class="headerlink" href="#advanced-user-guide-sangerread-ab1" title="Permalink to this headline">¶</a></h1>
<p><em>SangerRead</em> is in the bottommost level of sangeranalyseR (<a class="reference internal" href="#sangerread-hierarchy"><span class="std std-ref">Figure_1</span></a>), and each <em>SangerRead</em> object corresponds to a single read (one <strong>AB1</strong> file) in a Sanger sequencing experiment. <em>SangerRead</em> class extends <em>sangerseq</em> class from <a class="reference external" href="https://www.bioconductor.org/packages/release/bioc/html/sangerseqR.html">sangerseqR</a> package and contains input parameters and results of quality trimming and chromatogram. In this section, we are going to go through detailed sangeranalyseR data analysis steps in <em>SangerRead level</em> with <strong>AB1</strong> file input.</p>
<div class="figure align-center" id="id6">
<span id="sangerread-hierarchy"></span><a class="reference internal image-reference" href="../_images/SangerRead_hierarchy.png"><img alt="../_images/SangerRead_hierarchy.png" src="../_images/SangerRead_hierarchy.png" style="width: 626.6px; height: 217.8px;" /></a>
<p class="caption"><span class="caption-text">Figure 1. Hierarchy of classes in sangeranalyseR, <em>SangerRead</em> level.</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
<div class="section" id="preparing-sangerread-ab1-input">
<h2>Preparing <em>SangerRead</em> <strong>AB1</strong> input<a class="headerlink" href="#preparing-sangerread-ab1-input" title="Permalink to this headline">¶</a></h2>
<p>The main input file format to create <em>SangerRead</em> instance is <strong>AB1</strong>. Before starting the analysis, users need to prepare one target <strong>AB1</strong> file. The only hard regulation of the filename is that the input file must have <strong>.ab1</strong> as its file extension. There are some suggestions about the filename in the note below:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><strong>AB1</strong> file should be indexed for better consistency with file-naming regulation for <em>SangerContig</em> and <em>SangerAlignment</em>.</p></li>
<li><p>Forward or reverse direction should be specified in the filename.</p></li>
</ul>
</div>
<p><a class="reference internal" href="#sangerread-file-structure"><span class="std std-ref">Figure_2</span></a> shows the suggested file-naming strategy. The filename should contain four main parts: <strong>&quot;Contig name&quot;</strong>, <strong>&quot;Index number&quot;</strong>, <strong>&quot;Direction&quot;</strong> and <strong>&quot;ab1 file extension&quot;</strong>.</p>
<ul class="simple">
<li><p><strong>&quot;Contig name&quot;</strong> :  <code class="code docutils literal notranslate"><span class="pre">Achl_RBNII397-13</span></code></p></li>
<li><p><strong>&quot;Index number&quot;</strong> :  <code class="code docutils literal notranslate"><span class="pre">1</span></code></p></li>
<li><p><strong>&quot;Direction&quot;</strong> :  <code class="code docutils literal notranslate"><span class="pre">F</span></code></p></li>
<li><p><strong>&quot;ab1 file extension&quot;</strong> :  <code class="code docutils literal notranslate"><span class="pre">.ab1</span></code></p></li>
</ul>
<div class="figure align-center" id="id7">
<span id="sangerread-file-structure"></span><a class="reference internal image-reference" href="../_images/SangerRead_file_structure.png"><img alt="../_images/SangerRead_file_structure.png" src="../_images/SangerRead_file_structure.png" style="width: 248.0px; height: 33.6px;" /></a>
<p class="caption"><span class="caption-text">Figure 2. <em>SangerRead</em> filename regulation.</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
<p>In <em>SangerRead</em> section, it is not compulsory to follow the file-naming regulation because users can directly specify the filename in input (see <a class="reference internal" href="#creating-sangerread-instance-from-ab1"><span class="std std-ref">Creating SangerRead instance from AB1</span></a>); however, in the <em>SangerContig</em> and <em>SangerAlignment</em>, sangeranalyseR will automatically group files, so it is compulsory to have systematic file-naming strategy. For more details, please read <a class="reference internal" href="advance_user_SangerContig_ab1.html#advanced-user-guide-sangercontig-ab1"><span class="std std-ref">Advanced User Guide - SangerContig (AB1)</span></a> and <a class="reference internal" href="advance_user_SangerAlignment_ab1.html#advanced-user-guide-sangeralignment-ab1"><span class="std std-ref">Advanced User Guide - SangerAlignment (AB1)</span></a>. <a class="reference internal" href="#sangeranalyser-filename-convention-sangerread"><span class="std std-ref">Figure_3</span></a> shows the suggested <strong>AB1</strong> file-naming regulation.</p>
<div class="figure align-center" id="id8">
<span id="sangeranalyser-filename-convention-sangerread"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_filename_convention.png"><img alt="../_images/sangeranalyseR_filename_convention.png" src="../_images/sangeranalyseR_filename_convention.png" style="width: 621.0px; height: 28.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 3. Suggested <strong>AB1</strong> file-naming regulation - <em>SangerRead</em>.</span><a class="headerlink" href="#id8" title="Permalink to this image">¶</a></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="creating-sangerread-instance-from-ab1">
<h2>Creating <em>SangerRead</em> instance from <strong>AB1</strong><a class="headerlink" href="#creating-sangerread-instance-from-ab1" title="Permalink to this headline">¶</a></h2>
<p>After preparing the <em>SangerRead</em> input <strong>AB1</strong> file, the next step is to create the <em>SangerRead</em> instance by running the <code class="code docutils literal notranslate"><span class="pre">SangerRead</span></code> constructor function or <code class="code docutils literal notranslate"><span class="pre">new</span></code> method. The constructor function is a wrapper for the <code class="code docutils literal notranslate"><span class="pre">new</span></code> method which makes instance creation more intuitive. The inputs include <strong>Basic Parameters</strong>, <strong>Trimming Parameters</strong>, and <strong>Chromatogram Parameters</strong>, and all of them have default values. In the example below, we show both <em>SangerRead</em> creation methods with important parameters.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerRead instance</span>
<span class="n">sangerReadF</span> <span class="o">&lt;-</span> <span class="nf">SangerRead</span><span class="p">(</span><span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                          <span class="n">readFeature</span>           <span class="o">=</span> <span class="s">&quot;Forward Read&quot;</span><span class="p">,</span>
                          <span class="n">readFileName</span>          <span class="o">=</span> <span class="s">&quot;Achl_RBNII397-13_1_F.ab1&quot;</span><span class="p">,</span>
                          <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                          <span class="n">TrimmingMethod</span>        <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                          <span class="n">M1TrimmingCutoff</span>      <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                          <span class="n">M2CutoffQualityScore</span>  <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                          <span class="n">M2SlidingWindowSize</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                          <span class="n">baseNumPerRow</span>         <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                          <span class="n">heightPerRow</span>          <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                          <span class="n">signalRatioCutoff</span>     <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                          <span class="n">showTrimmed</span>           <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># using `new` method to create SangerRead instance</span>
<span class="n">sangerReadF</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerRead&quot;</span><span class="p">,</span>
                    <span class="n">inputSource</span>           <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                    <span class="n">readFeature</span>           <span class="o">=</span> <span class="s">&quot;Forward Read&quot;</span><span class="p">,</span>
                    <span class="n">readFileName</span>          <span class="o">=</span> <span class="s">&quot;Achl_RBNII397-13_1_F.ab1&quot;</span><span class="p">,</span>
                    <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                    <span class="n">TrimmingMethod</span>        <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                    <span class="n">M1TrimmingCutoff</span>      <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                    <span class="n">M2CutoffQualityScore</span>  <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                    <span class="n">M2SlidingWindowSize</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                    <span class="n">baseNumPerRow</span>         <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                    <span class="n">heightPerRow</span>          <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                    <span class="n">signalRatioCutoff</span>     <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                    <span class="n">showTrimmed</span>           <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>The inputs of <code class="code docutils literal notranslate"><span class="pre">SangerRead</span></code> constructor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method are the same. For more details about <em>SangerRead</em> inputs and slots definition, please refer to the <a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/sangeranalyseR/man/sangeranalyseR.pdf">sangeranalyseR reference manual</a>. The created <em>SangerRead</em> instance, <code class="code docutils literal notranslate"><span class="pre">sangerReadF</span></code>, is used as the input for the following functions.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="visualizing-sangerread-trimmed-read">
<h2>Visualizing <em>SangerRead</em> trimmed read<a class="headerlink" href="#visualizing-sangerread-trimmed-read" title="Permalink to this headline">¶</a></h2>
<p>Before going to <a class="reference internal" href="#writing-sangerread-fasta-file-ab1"><span class="std std-ref">Writing SangerRead FASTA file (AB1)</span></a> and <a class="reference internal" href="#generating-sangerread-report-ab1"><span class="std std-ref">Generating SangerRead report (AB1)</span></a> pages, it is suggested to visualize the trimmed <em>SangerRead</em>. Run the <code class="code docutils literal notranslate"><span class="pre">qualityBasePlot</span></code> function to get the result in <a class="reference internal" href="#sangerread-qualitybaseplot"><span class="std std-ref">Figure_4</span></a>. It shows the quality score for each base pairs and the trimming start/end points of the sequence.</p>
<div class="figure align-center" id="id9">
<span id="sangerread-qualitybaseplot"></span><a class="reference internal image-reference" href="../_images/SangerRead_qualityBasePlot.png"><img alt="../_images/SangerRead_qualityBasePlot.png" src="../_images/SangerRead_qualityBasePlot.png" style="width: 498.0px; height: 337.2px;" /></a>
<p class="caption"><span class="caption-text">Figure 4. <em>SangerRead</em> trimmed read visualization.</span><a class="headerlink" href="#id9" title="Permalink to this image">¶</a></p>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">qualityBasePlot</span><span class="p">(</span><span class="n">sangerReadF</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="updating-sangerread-quality-trimming-parameters">
<h2>Updating <em>SangerRead</em> quality trimming parameters<a class="headerlink" href="#updating-sangerread-quality-trimming-parameters" title="Permalink to this headline">¶</a></h2>
<p>In the previous <a class="reference internal" href="#creating-sangerread-instance-from-ab1"><span class="std std-ref">Creating SangerRead instance from AB1</span></a> part, the constructor function applies the quality trimming parameters to the read. These parameters are not fixed. After instance creation, users can run <code class="code docutils literal notranslate"><span class="pre">updateQualityParam</span></code> function which will change the <em>QualityReport</em> instance inside the <em>SangerRead</em> and update frameshift amino acid sequences.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">newSangerRead</span> <span class="o">&lt;-</span> <span class="nf">updateQualityParam</span><span class="p">(</span><span class="n">sangerReadF</span><span class="p">,</span>
                                    <span class="n">TrimmingMethod</span>       <span class="o">=</span> <span class="s">&quot;M2&quot;</span><span class="p">,</span>
                                    <span class="n">M1TrimmingCutoff</span>     <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                    <span class="n">M2CutoffQualityScore</span> <span class="o">=</span> <span class="m">29</span><span class="p">,</span>
                                    <span class="n">M2SlidingWindowSize</span>  <span class="o">=</span> <span class="m">15</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="writing-sangerread-fasta-file-ab1">
<h2>Writing <em>SangerRead</em> FASTA file <sub>(AB1)</sub><a class="headerlink" href="#writing-sangerread-fasta-file-ab1" title="Permalink to this headline">¶</a></h2>
<p>After quality trimming, users can write <code class="code docutils literal notranslate"><span class="pre">sangerReadF</span></code> into a <strong>FASTA</strong> file. Below is the one-liner that needs to be run. This function, <code class="code docutils literal notranslate"><span class="pre">writeFasta</span></code>, mainly depends on <code class="code docutils literal notranslate"><span class="pre">writeXStringSet</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> R package. Users can further set the compression level through it.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">newSangerRead</span><span class="p">,</span>
           <span class="n">outputDir</span>         <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
           <span class="n">compress</span>          <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
           <span class="n">compression_level</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can download the <a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerRead/AB1/ACHLO006-09[LCO1490_t1,HCO2198_t1]_1_F.fa">output FASTA file</a> of this example.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="generating-sangerread-report-ab1">
<h2>Generating <em>SangerRead</em> report <sub>(AB1)</sub><a class="headerlink" href="#generating-sangerread-report-ab1" title="Permalink to this headline">¶</a></h2>
<p>Last but not least, users can save <code class="code docutils literal notranslate"><span class="pre">sangerReadF</span></code> into a static <strong>HTML</strong> report by knitting <strong>Rmd</strong> files. In this example, <code class="code docutils literal notranslate"><span class="pre">tempdir</span></code> function will generate a random path.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">newSangerRead</span><span class="p">,</span>
               <span class="n">outputDir</span> <span class="o">=</span> <span class="nf">tempdir</span><span class="p">())</span>
</pre></div>
</div>
<p><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerRead/AB1/ACHLO006-09[LCO1490_t1,HCO2198_t1]_1_F/SangerRead_Report_ab1.html">SangerRead_Report_ab1.html</a> is the generated <em>SangerRead</em> report html of this example. Users can access to '<em>Basic Information</em>', '<em>DNA Sequence</em>', '<em>Amino Acids Sequence</em>', '<em>Quality Trimming</em>' and '<em>Chromatogram</em>' sections inside this report.</p>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="a-reproducible-example-sangerread-ab1">
<h2>A Reproducible Example (<em>SangerRead</em>, <strong>ab1</strong>)<a class="headerlink" href="#a-reproducible-example-sangerread-ab1" title="Permalink to this headline">¶</a></h2>
<div class="section" id="id1">
<h3>1. Preparing <em>SangerRead</em> <strong>AB1</strong> input<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<p>The data of this example is in the sangeranalyseR package; thus, you can simply get its path from the library.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">inputFilesPath</span> <span class="o">&lt;-</span> <span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata/&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s">&quot;sangeranalyseR&quot;</span><span class="p">)</span>
<span class="n">A_chloroticaFFN</span> <span class="o">&lt;-</span> <span class="nf">file.path</span><span class="p">(</span><span class="n">inputFilesPath</span><span class="p">,</span>
                             <span class="s">&quot;Allolobophora_chlorotica&quot;</span><span class="p">,</span>
                             <span class="s">&quot;ACHLO&quot;</span><span class="p">,</span>
                             <span class="s">&quot;Achl_ACHLO006-09_1_F.ab1&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id2">
<h3>2. Creating <em>SangerRead</em> instance from <strong>AB1</strong><a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<p>Run the following on-liner, SangerRead <code class="code docutils literal notranslate"><span class="pre">constructor</span></code> or <code class="code docutils literal notranslate"><span class="pre">new</span></code> method, to create the <em>SangerRead</em> object.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># using `constructor` function to create SangerRead instance</span>
<span class="n">sangerReadF</span> <span class="o">&lt;-</span> <span class="nf">SangerRead</span><span class="p">(</span><span class="n">readFeature</span>           <span class="o">=</span> <span class="s">&quot;Forward Read&quot;</span><span class="p">,</span>
                          <span class="n">readFileName</span>          <span class="o">=</span> <span class="n">A_chloroticaFFN</span><span class="p">)</span>

<span class="c1"># using `new` method to create SangerRead instance</span>
<span class="n">sangerReadF</span> <span class="o">&lt;-</span> <span class="nf">new</span><span class="p">(</span><span class="s">&quot;SangerRead&quot;</span><span class="p">,</span>
                   <span class="n">readFeature</span>           <span class="o">=</span> <span class="s">&quot;Forward Read&quot;</span><span class="p">,</span>
                   <span class="n">readFileName</span>          <span class="o">=</span> <span class="n">A_chloroticaFFN</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>INFO [2021-29-06 16:28:39] ------------------------------------------------
INFO [2021-29-06 16:28:39] -------- Creating &#39;SangerRead&#39; instance --------
INFO [2021-29-06 16:28:39] ------------------------------------------------
INFO [2021-29-06 16:28:39] &gt;&gt; Forward Read: Creating abif &amp; sangerseq ...
INFO [2021-29-06 16:28:39]     &gt;&gt; Creating Forward Read raw abif ...
INFO [2021-29-06 16:28:39]     &gt;&gt; Creating Forward Read raw sangerseq ...
INFO [2021-29-06 16:28:39]           * Making basecall !!
INFO [2021-29-06 16:28:40]           * Updating slots in &#39;SangerRead&#39; instance !!
SUCCESS [2021-29-06 16:28:40] --------------------------------------------------------
SUCCESS [2021-29-06 16:28:40] -------- &#39;SangerRead&#39; S4 instance is created !! --------
SUCCESS [2021-29-06 16:28:40] --------------------------------------------------------
SUCCESS [2021-29-06 16:28:40]    &gt;&gt; &#39;Achl_ACHLO006-09_1_F.ab1&#39; is created (Forward Read; ABIF).
INFO [2021-29-06 16:28:40]    &gt;&gt; Read is trimmed by &#39;M1 - Mott’s trimming algorithm&#39;.
DEBUG [2021-29-06 16:28:40]    &gt;&gt; For more information, please run &#39;object&#39;.
DEBUG [2021-29-06 16:28:40]    &gt;&gt; Run &#39;object@objectResults@readResultTable&#39; to check the result of the Sanger read
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id3">
<h3>3. Visualizing <em>SangerRead</em> trimmed read<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>Launch an interactive plotly plot to check the trimmed read.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">qualityBasePlot</span><span class="p">(</span><span class="n">sangerReadF</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id4">
<h3>4. Writing <em>SangerRead</em> FASTA file <sub>(AB1)</sub><a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>Write the trimmed read into a FASTA file.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">sangerReadF</span><span class="p">)</span>
</pre></div>
</div>
<div class="toggle docutils container">
<div class="header docutils container">
<p>Following is the R shell output that you will get.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">16</span><span class="p">:</span><span class="mi">30</span><span class="p">:</span><span class="mi">17</span><span class="p">]</span> <span class="n">Your</span> <span class="nb">input</span> <span class="ow">is</span> <span class="s1">&#39;SangerRead&#39;</span> <span class="n">S4</span> <span class="n">instance</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">16</span><span class="p">:</span><span class="mi">30</span><span class="p">:</span><span class="mi">17</span><span class="p">]</span> <span class="o">&gt;&gt;&gt;</span> <span class="n">outputDir</span> <span class="p">:</span> <span class="o">/</span><span class="n">private</span><span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">33</span><span class="o">/</span><span class="mi">7</span><span class="n">v38jdjd2874jcxb6l71m00h0000gn</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">RtmpRAPaMV</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">16</span><span class="p">:</span><span class="mi">30</span><span class="p">:</span><span class="mi">17</span><span class="p">]</span> <span class="n">Start</span> <span class="n">writing</span> <span class="s1">&#39;/Library/Frameworks/R.framework/Versions/4.0/Resources/library/sangeranalyseR/extdata//Allolobophora_chlorotica/ACHLO/Achl_ACHLO006-09_1_F.ab1&#39;</span> <span class="n">to</span> <span class="n">FASTA</span> <span class="nb">format</span> <span class="o">...</span>
<span class="n">INFO</span> <span class="p">[</span><span class="mi">2021</span><span class="o">-</span><span class="mi">29</span><span class="o">-</span><span class="mi">06</span> <span class="mi">16</span><span class="p">:</span><span class="mi">30</span><span class="p">:</span><span class="mi">17</span><span class="p">]</span> <span class="o">&gt;&gt;</span> <span class="s1">&#39;/private/var/folders/33/7v38jdjd2874jcxb6l71m00h0000gn/T/RtmpRAPaMV/Achl_ACHLO006-09_1_F.fa&#39;</span> <span class="ow">is</span> <span class="n">written</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;/private/var/folders/33/7v38jdjd2874jcxb6l71m00h0000gn/T/RtmpRAPaMV/Achl_ACHLO006-09_1_F.fa&quot;</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>And you will get one FASTA file:</p>
<ol class="arabic simple">
<li><p><a class="reference download internal" download="" href="../_downloads/789dce984eb69dd506123d3e0ec07b0b/Achl_ACHLO006-09_1_F.fa"><code class="xref download docutils literal notranslate"><span class="pre">Achl_ACHLO006-09_1_F.fa</span></code></a></p></li>
</ol>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="id5">
<h3>5. Generating <em>SangerRead</em> report <sub>(AB1)</sub><a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>Last but not least, generate an Rmarkdown report to store all sequence information.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">sangerReadF</span><span class="p">)</span>
</pre></div>
</div>
<hr class="docutils" />
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="advance_user_SangerContig_ab1.html" class="btn btn-neutral float-right" title="Advanced User Guide - SangerContig (AB1)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="beginner.html" class="btn btn-neutral float-left" title="Beginners Guide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2019, Kuan-Hao Chao.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 <script type="text/javascript">
    $(document).ready(function() {
        $(".toggle > *").hide();
        $(".toggle .header").show();
        $(".toggle .header").click(function() {
            $(this).parent().children().not(".header").toggle(400);
            $(this).parent().children(".header").toggleClass("open");
        })
    });
</script>


</body>
</html>