

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="English" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="English" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Advanced User Guide - SangerAlignment (AB1) &mdash; sangeranalyseR  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced User Guide - SangerRead (FASTA)" href="advance_user_SangerRead_fasta.html" />
    <link rel="prev" title="Advanced User Guide - SangerContig (AB1)" href="advance_user_SangerContig_ab1.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> sangeranalyseR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner.html">Beginners Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_ab1.html">Advanced User Guide - <em>SangerRead</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_ab1.html">Advanced User Guide - <em>SangerContig</em> (<strong>AB1</strong>)</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preparing-sangeralignment-ab1-input">Preparing <em>SangerAlignment</em> <strong>AB1</strong> input</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-sangeralignment-instance-from-ab1">Creating <em>SangerAlignment</em> instance from <strong>AB1</strong></a></li>
<li class="toctree-l2"><a class="reference internal" href="#updating-sangeralignment-quality-trimming-parameters">Updating <em>SangerAlignment</em> quality trimming parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#launching-sangeralignment-shiny-app">Launching <em>SangerAlignment</em> Shiny app</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sangeralignment-page-sa-app"><em>SangerAlignment</em> page (SA app)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sangercontig-page-sa-app"><em>SangerContig</em> page (SA app)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sangerread-page-sa-app"><em>SangerRead</em> page (SA app)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#writing-sangeralignment-fasta-files-ab1">Writing <em>SangerAlignment</em> FASTA files <sub>(AB1)</sub></a></li>
<li class="toctree-l2"><a class="reference internal" href="#generating-sangeralignment-report-ab1">Generating <em>SangerAlignment</em> report <sub>(AB1)</sub></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerRead_fasta.html">Advanced User Guide - <em>SangerRead</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerContig_fasta.html">Advanced User Guide - <em>SangerContig</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="advance_user_SangerAlignment_fasta.html">Advanced User Guide - <em>SangerAlignment</em> (<strong>FASTA</strong>)</a></li>
<li class="toctree-l1"><a class="reference internal" href="conclusion.html">Conclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="contact.html">Contact</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">sangeranalyseR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/content/advance_user_SangerAlignment_ab1.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="advanced-user-guide-sangeralignment-ab1">
<h1>Advanced User Guide - <em>SangerAlignment</em> (<strong>AB1</strong>)<a class="headerlink" href="#advanced-user-guide-sangeralignment-ab1" title="Permalink to this headline">¶</a></h1>
<p><em>SangerAlignment</em> is the highest class level in sangeranalyseR showed in <a class="reference internal" href="#sangeralignment-hierachy"><span class="std std-ref">Figure_1</span></a>. It contains <em>SangerContig</em> list and the contigs alignment result. Users can access to <em>SangerContig</em> and <em>SangerRead</em> instance inside <em>SangerAlignment</em> instance. In this section, we are going to go through detailed sangeranalyseR data analysis steps in <em>SangerAlignment</em> level from <strong>AB1</strong> file input.</p>
<div class="figure align-center" id="id1">
<span id="sangeralignment-hierachy"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_hierachy.png"><img alt="../_images/SangerAlignment_hierachy.png" src="../_images/SangerAlignment_hierachy.png" style="width: 545.4px; height: 218.8px;" /></a>
<p class="caption"><span class="caption-text">Figure 1. Classes hierarchy in sangeranalyseR, <em>SangerAlignment</em> level.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<div class="section" id="preparing-sangeralignment-ab1-input">
<h2>Preparing <em>SangerAlignment</em> <strong>AB1</strong> input<a class="headerlink" href="#preparing-sangeralignment-ab1-input" title="Permalink to this headline">¶</a></h2>
<p>The main input file format to create <em>SangerAlignment</em> instance is <strong>AB1</strong>. Before starting the analysis, users need to prepare a directory which contains all the <strong>AB1</strong> files. Here are some filename regulations:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>All the input files must have <strong>.ab1</strong> as its file extension</p></li>
<li><p>The reads that belong to the same contig must have the same contig name in its filename.</p></li>
<li><p>Forward or reverse direction also has to be specified in the filename.</p></li>
</ul>
</div>
<p>There are three parameters, <code class="code docutils literal notranslate"><span class="pre">parentDirectory</span></code>, <code class="code docutils literal notranslate"><span class="pre">suffixForwardRegExp</span></code>, and <code class="code docutils literal notranslate"><span class="pre">suffixReverseRegExp</span></code>, that users need to provide so that program can automatically group all <strong>AB1</strong> files.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">parentDirectory</span></code>: The root directory that contains all the <strong>AB1</strong> files. It can be absolute or relative path. We suggest users to put only target <strong>AB1</strong> files inside this directory without other unrelated files.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">suffixForwardRegExp</span></code>: The value of this parameter is a regular expression that matches all filenames in forward direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used to select forward reads from all <strong>AB1</strong> files.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">suffixReverseRegExp</span></code>: The value of this parameter is a regular expression that matches all filenames in reverse direction. <code class="code docutils literal notranslate"><span class="pre">grepl</span></code> function in R is used to select reverse reads from all <strong>AB1</strong> files.</p></li>
</ul>
</div>
<p>For basic input files preparation example, please go to <a class="reference internal" href="beginner.html#beginners-guide"><span class="std std-ref">Beginners Guide</span></a>. Here, we have another more complicated example.</p>
<div class="figure align-center" id="id2">
<span id="sangeralignment-file-structure-complex"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_file_structure_complex.png"><img alt="../_images/SangerAlignment_file_structure_complex.png" src="../_images/SangerAlignment_file_structure_complex.png" style="width: 435.4px; height: 826.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 2. Input ab1 files inside the parent directory, <code class="code docutils literal notranslate"><span class="pre">./tmp/</span></code>.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
<p><a class="reference internal" href="#sangeralignment-file-structure-complex"><span class="std std-ref">Figure_2</span></a> shows the file naming regulation and directory hierarchy. In this example, the parent directory is <code class="code docutils literal notranslate"><span class="pre">extdata</span></code> and the directories in first layer are <code class="code docutils literal notranslate"><span class="pre">Allolobophora_chlorotica</span></code> and <code class="code docutils literal notranslate"><span class="pre">Drosophila_melanogaster</span></code>. All target <strong>AB1</strong> files need to be inside parent directory but it is not necessary to put them in the same level. sangeranalyseR will recursively search all files with <strong>.ab1</strong> file extension and automatically group reads with the same contig name. The direction of reads in each contig will be grouped by matching <code class="code docutils literal notranslate"><span class="pre">suffixForwardRegExp</span></code> and <code class="code docutils literal notranslate"><span class="pre">suffixReverseRegExp</span></code> with filenames. Therefore, it is important to carefully select <code class="code docutils literal notranslate"><span class="pre">suffixForwardRegExp</span></code> and <code class="code docutils literal notranslate"><span class="pre">suffixReverseRegExp</span></code>. The bad file naming regulation and wrong regex matching might accidentally include reverse reads into the forward read list or vice versa, which will make the program generate totally wrong results. Therefore, users should have a consistent naming strategy. In this example, &quot;<code class="code docutils literal notranslate"><span class="pre">_[0-9]+_F</span></code>&quot;, &quot;<code class="code docutils literal notranslate"><span class="pre">_[0-9]+_R</span></code>&quot; for matching forward and reverse reads are highly suggested and are used as default. It is a good habit to index your reads in the same contig group because there might be more than one read that are in the forward or reverse direction.</p>
<div class="figure align-center" id="id3">
<span id="sangeranalyser-filename-convention-sangeralignment"></span><a class="reference internal image-reference" href="../_images/sangeranalyseR_filename_convention.png"><img alt="../_images/sangeranalyseR_filename_convention.png" src="../_images/sangeranalyseR_filename_convention.png" style="width: 621.0px; height: 28.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 3. Suggested <strong>AB1</strong> file naming regulation - <em>SangerAlignment</em>.</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
<p><a class="reference internal" href="#sangeranalyser-filename-convention-sangeralignment"><span class="std std-ref">Figure_3</span></a> shows the suggested <strong>AB1</strong> file naming regulation. Users are strongly recommended to follow this file naming regulation and use the default <code class="code docutils literal notranslate"><span class="pre">suffixForwardRegExp</span></code> : &quot;<code class="code docutils literal notranslate"><span class="pre">_[0-9]+_F</span></code>&quot; and <code class="code docutils literal notranslate"><span class="pre">suffixReverseRegExp</span></code> : &quot;<code class="code docutils literal notranslate"><span class="pre">_[0-9]+_R</span></code>&quot; to reduce any chance of error.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="creating-sangeralignment-instance-from-ab1">
<h2>Creating <em>SangerAlignment</em> instance from <strong>AB1</strong><a class="headerlink" href="#creating-sangeralignment-instance-from-ab1" title="Permalink to this headline">¶</a></h2>
<p>After preparing the input directory, we can create the <em>SangerAlignment</em> S4 instance by running <code class="code docutils literal notranslate"><span class="pre">SangerAlignment</span></code> constructor function or <code class="code docutils literal notranslate"><span class="pre">new</span></code> method. The constructor function is a wrapper for <code class="code docutils literal notranslate"><span class="pre">new</span></code> method and it makes instance creation more intuitive. Most parameters in the constructor have their own default values. In the constructor below, we list important parameters. For a simpler command, please go to <a class="reference internal" href="quickstart.html#quick-start-guide"><span class="std std-ref">Quick Start Guide</span></a>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">sangerAlignment</span> <span class="o">&lt;-</span> <span class="nf">SangerAlignment</span><span class="p">(</span><span class="n">inputSource</span>          <span class="o">=</span> <span class="s">&quot;ABIF&quot;</span><span class="p">,</span>
                                   <span class="n">parentDirectory</span>      <span class="o">=</span> <span class="s">&quot;./tmp/&quot;</span><span class="p">,</span>
                                   <span class="n">suffixForwardRegExp</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]+_F&quot;</span><span class="p">,</span>
                                   <span class="n">suffixReverseRegExp</span>  <span class="o">=</span> <span class="s">&quot;_[0-9]+_R&quot;</span><span class="p">,</span>
                                   <span class="n">refAminoAcidSeq</span>      <span class="o">=</span> <span class="s">&quot;SRQWLFSTNHKDIGTLYFIFGAWAGMVGTSLSILIRAELGHPGALIGDDQIYNVIVTAHAFIMIFFMVMPIMIGGFGNWLVPLMLGAPDMAFPRMNNMSFWLLPPALSLLLVSSMVENGAGTGWTVYPPLSAGIAHGGASVDLAIFSLHLAGISSILGAVNFITTVINMRSTGISLDRMPLFVWSVVITALLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGSLGMIYAMLAIGLLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQLSYSPAILWALGFVFLFTVGGLTGVVLANSSVDIILHDTYYVVAHFHYVLSMGAVFAIMAGFIHWYPLFTGLTLNNKWLKSHFIIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDAYTTWNIVSTIGSTISLLGILFFFFIIWESLVSQRQVIYPIQLNSSIEWYQNTPPAEHSYSELPLLTN&quot;</span><span class="p">,</span>
                                   <span class="n">TrimmingMethod</span>        <span class="o">=</span> <span class="s">&quot;M1&quot;</span><span class="p">,</span>
                                   <span class="n">M1TrimmingCutoff</span>      <span class="o">=</span> <span class="m">0.0001</span><span class="p">,</span>
                                   <span class="n">M2CutoffQualityScore</span>  <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                   <span class="n">M2SlidingWindowSize</span>   <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                   <span class="n">baseNumPerRow</span>         <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
                                   <span class="n">heightPerRow</span>          <span class="o">=</span> <span class="m">200</span><span class="p">,</span>
                                   <span class="n">signalRatioCutoff</span>     <span class="o">=</span> <span class="m">0.33</span><span class="p">,</span>
                                   <span class="n">showTrimmed</span>           <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                   <span class="n">minReadsNum</span>           <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
                                   <span class="n">minReadLength</span>         <span class="o">=</span> <span class="m">20</span><span class="p">,</span>
                                   <span class="n">minFractionCall</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                   <span class="n">maxFractionLost</span>       <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                   <span class="n">geneticCode</span>           <span class="o">=</span> <span class="n">GENETIC_CODE</span><span class="p">,</span>
                                   <span class="n">acceptStopCodons</span>      <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
                                   <span class="n">readingFrame</span>          <span class="o">=</span> <span class="m">1</span><span class="p">,</span>
                                   <span class="n">minFractionCallSA</span>     <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                   <span class="n">maxFractionLostSA</span>     <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
                                   <span class="n">processorsNum</span>         <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span>
</pre></div>
</div>
<p>The inputs of <code class="code docutils literal notranslate"><span class="pre">SangerAlignment</span></code> constructor function and <code class="code docutils literal notranslate"><span class="pre">new</span></code> method are same. For more details about <em>SangerAlignment</em> inputs and slots definition, please refer to <a class="reference external" href="http://packages.python.org/an_example_pypi_project/">sangeranalyseR reference manual (need update)</a>.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="updating-sangeralignment-quality-trimming-parameters">
<h2>Updating <em>SangerAlignment</em> quality trimming parameters<a class="headerlink" href="#updating-sangeralignment-quality-trimming-parameters" title="Permalink to this headline">¶</a></h2>
<p>In the previous <a class="reference internal" href="#creating-sangeralignment-instance-from-ab1"><span class="std std-ref">Creating SangerAlignment instance from AB1</span></a> part, the constructor function will apply the quality trimming parameters to all reads. After creating the <em>SangerAlignment</em> S4 instance, users can change the trimming parameters by running <code class="code docutils literal notranslate"><span class="pre">updateQualityParam</span></code> function which will update all reads with the new trimming parameters and redo reads alignment in <em>SangerContig</em> and contigs alignment in <em>SangerAlignment</em>. If users want to do quality trimming read by read instead all at once, please read <a class="reference internal" href="#launching-sangeralignment-shiny-app"><span class="std std-ref">Launching SangerAlignment Shiny app</span></a>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">newSangerAlignment</span> <span class="o">&lt;-</span> <span class="nf">updateQualityParam</span><span class="p">(</span><span class="n">sangerAlignment</span><span class="p">,</span>
                                         <span class="n">TrimmingMethod</span>       <span class="o">=</span> <span class="s">&quot;M2&quot;</span><span class="p">,</span>
                                         <span class="n">M1TrimmingCutoff</span>     <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
                                         <span class="n">M2CutoffQualityScore</span> <span class="o">=</span> <span class="m">29</span><span class="p">,</span>
                                         <span class="n">M2SlidingWindowSize</span>  <span class="o">=</span> <span class="m">15</span><span class="p">)</span>
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="launching-sangeralignment-shiny-app">
<h2>Launching <em>SangerAlignment</em> Shiny app<a class="headerlink" href="#launching-sangeralignment-shiny-app" title="Permalink to this headline">¶</a></h2>
<p>We create an interactive local Shiny app for users to go into each <em>SangerRead</em> and <em>SangerContig</em> in <em>SangerAlignment</em> instance. Users only need to run one function with previously created instance as input and the <em>SangerAlignment</em> Shiny app will pop up. Here, we will go through pages in the three levels.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">launchApp</span><span class="p">(</span><span class="n">sangerAlignment</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="sangeralignment-page-sa-app">
<h3><em>SangerAlignment</em> page (SA app)<a class="headerlink" href="#sangeralignment-page-sa-app" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="#sangeralignment-shinyapp-1"><span class="std std-ref">Figure 4</span></a> is the initial page and the toppest layer of <em>SangerAlignment</em> App. It provides basic parameters in <em>SangerAlignment</em> instance, contigs alignment result and phylogenetic tree etc. Before checking the results, users need to click “Re-calculate Contigs Alignment” button to do contigs alignment in order to get the updated results. From the left-hand side panel, we can clearly see the hierarchy of the <em>SangerAlignment</em> S4 instance and easily access to all reads and contigs in it.</p>
<div class="figure align-center" id="id4">
<span id="sangeralignment-shinyapp-1"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_1.png"><img alt="../_images/SangerAlignment_ShinyApp_1.png" src="../_images/SangerAlignment_ShinyApp_1.png" style="width: 768.0px; height: 417.6px;" /></a>
<p class="caption"><span class="caption-text">Figure 4. <em>SangerAlignment</em> Shiny app initial page - <em>SangerAlignment</em> Page.</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
<p>Scroll down a bit, users can see the contigs alignment result generated by <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> R package embedded in <em>SangerAlignment</em> page. <a class="reference internal" href="#sangeralignment-shinyapp-2"><span class="std std-ref">Figure 5</span></a> shows the contigs alignment result.</p>
<div class="figure align-center" id="id5">
<span id="sangeralignment-shinyapp-2"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_2.png"><img alt="../_images/SangerAlignment_ShinyApp_2.png" src="../_images/SangerAlignment_ShinyApp_2.png" style="width: 765.0px; height: 382.5px;" /></a>
<p class="caption"><span class="caption-text">Figure 5. <em>SangerAlignment</em> Page - contigs alignment result.</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
<p>In <em>SangerAlignment</em> page, the phylogenetic tree result is provided as well (<a class="reference internal" href="#sangeralignment-shinyapp-3"><span class="std std-ref">Figure 6</span></a>). The tree is generated by <a class="reference external" href="https://cran.r-project.org/web/packages/ape/index.html">ape</a> R package which uses neighbor-joining algorithm.</p>
<div class="figure align-center" id="id6">
<span id="sangeralignment-shinyapp-3"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_3.png"><img alt="../_images/SangerAlignment_ShinyApp_3.png" src="../_images/SangerAlignment_ShinyApp_3.png" style="width: 727.0px; height: 331.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 6. <em>SangerAlignment</em> Page - phylogenetic tree result.</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="sangercontig-page-sa-app">
<h3><em>SangerContig</em> page (SA app)<a class="headerlink" href="#sangercontig-page-sa-app" title="Permalink to this headline">¶</a></h3>
<p>Now, let's go to the page in the next level, <em>SangerContig</em> page. Users can click into all contigs and check their results. <a class="reference internal" href="#sangeralignment-shinyapp-5"><span class="std std-ref">Figure 7</span></a> shows the overview page of Contig 1. Notice that there is a red “Re-calculate Contig” button. After changing the quality trimming parameters, users need to click the button before checking the results below in order to get the updated information.</p>
<div class="figure align-center" id="id7">
<span id="sangeralignment-shinyapp-5"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_5.png"><img alt="../_images/SangerAlignment_ShinyApp_5.png" src="../_images/SangerAlignment_ShinyApp_5.png" style="width: 710.4px; height: 386.28px;" /></a>
<p class="caption"><span class="caption-text">Figure 7. <em>SangerAlignment</em> Shiny app - <em>SangerContig</em> page.</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
<p>The information provided in this page includes : “input parameters”, “genetic code table”, “reference amino acid sequence”, “reads alignment”, “difference data frame”, “dendrogram”, “sample distance heatmap”, “indels data frame”, “stop codons data frame”. <a class="reference internal" href="#sangeralignment-shinyapp-6"><span class="std std-ref">Figure 8</span></a> and <a class="reference internal" href="#sangeralignment-shinyapp-7"><span class="std std-ref">Figure 9</span></a> show part of the results in the <em>SangerContig</em> page. The results are dynamic based on the trimming parameters from user inputs.</p>
<div class="figure align-center" id="id8">
<span id="sangeralignment-shinyapp-6"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_6.png"><img alt="../_images/SangerAlignment_ShinyApp_6.png" src="../_images/SangerAlignment_ShinyApp_6.png" style="width: 726.5px; height: 492.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 8. <em>SangerContig</em> page - contig-related parameters, genetic code and reference amino acid sequence.</span><a class="headerlink" href="#id8" title="Permalink to this image">¶</a></p>
</div>
<div class="figure align-center" id="id9">
<span id="sangeralignment-shinyapp-7"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_7.png"><img alt="../_images/SangerAlignment_ShinyApp_7.png" src="../_images/SangerAlignment_ShinyApp_7.png" style="width: 764.5px; height: 501.5px;" /></a>
<p class="caption"><span class="caption-text">Figure 9. <em>SangerContig</em> page - reads alignment and difference data frame.</span><a class="headerlink" href="#id9" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="sangerread-page-sa-app">
<h3><em>SangerRead</em> page (SA app)<a class="headerlink" href="#sangerread-page-sa-app" title="Permalink to this headline">¶</a></h3>
<p>Now, let's go to the page in the lowest level, <em>SangerRead</em> page. <em>SangerRead</em> page contains all details of a read including its trimming and chromatogram inputs and results. All reads are in &quot;forward&quot; or &quot;reverse&quot; direction. Under &quot;Contig Overview&quot; tab (<em>SangerContig</em> page), there are two expendable tabs, “Forward Reads” and “Reverse Reads” storing corresponding reads on the left-hand side navigation panel in <a class="reference internal" href="#sangeralignment-shinyapp-8"><span class="std std-ref">Figure 10</span></a>. In this example, there are one read in each tab and <a class="reference internal" href="#sangeralignment-shinyapp-8"><span class="std std-ref">Figure 10</span></a> shows the “1 - 1 Forward Read” page. It provides basic information, quality trimming inputs, chromatogram plotting inputs etc. Primary/secondary sequences in this figure are dynamic based on the <code class="code docutils literal notranslate"><span class="pre">signalRatioCutoff</span></code> value for base calling and the length of them are always same. Another thing to mention is that primary/secondary sequences and the sequences in the chromatogram in <a class="reference internal" href="#sangeralignment-shinyapp-14"><span class="std std-ref">Figure 15</span></a> below will always be same after trimming and their color codings for A/T/C/G are same as well.</p>
<div class="figure align-center" id="id10">
<span id="sangeralignment-shinyapp-8"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_8.png"><img alt="../_images/SangerAlignment_ShinyApp_8.png" src="../_images/SangerAlignment_ShinyApp_8.png" style="width: 672.0px; height: 364.34999999999997px;" /></a>
<p class="caption"><span class="caption-text">Figure 10. <em>SangerAlignment</em> Shiny app - <em>SangerRead</em> page.</span><a class="headerlink" href="#id10" title="Permalink to this image">¶</a></p>
</div>
<p>In quality trimming steps, we removes fragment at both ends of sequencing reads with low quality score. It is important because trimmed reads will improves alignment results. <a class="reference internal" href="#sangeralignment-shinyapp-9"><span class="std std-ref">Figure 11</span></a> shows the UI for Trimming Method 1 (M1): ‘Modified Mott Trimming’. This method is implemented in <a class="reference external" href="http://www.phrap.org/phredphrapconsed.html">Phred</a>. Users can change the cutoff score and click “Apply Trimming Parameters&quot; button to update the UI. The value of input must be between 0 and 1. If the input is invalid, the cutoff score will be set to default 0.0001.</p>
<div class="figure align-center" id="id11">
<span id="sangeralignment-shinyapp-9"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_9.png"><img alt="../_images/SangerAlignment_ShinyApp_9.png" src="../_images/SangerAlignment_ShinyApp_9.png" style="width: 693.0px; height: 324.0px;" /></a>
<p class="caption"><span class="caption-text">Figure 11. <em>SangerRead</em> page - Trimming Method 1 (M1): ‘Modified Mott Trimming’ UI.</span><a class="headerlink" href="#id11" title="Permalink to this image">¶</a></p>
</div>
<p><a class="reference internal" href="#sangeralignment-shinyapp-10"><span class="std std-ref">Figure 12</span></a> shows another quality trimming methods for users to choose from, Trimming Method 2 (M2): ‘Trimmomatics Sliding Window Trimming’. This method is implemented in <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatics</a>. Users can change the cutoff quality score as well as sliding window size and click “Apply Trimming Parameters&quot; button to update the UI. The value of cutoff quality score must be between 0 and 60 (default 20); the value of sliding window size must be between 0 and 40 (default 10). If the inputs are invalid, their values will be set to default.</p>
<div class="figure align-center" id="id12">
<span id="sangeralignment-shinyapp-10"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_10.png"><img alt="../_images/SangerAlignment_ShinyApp_10.png" src="../_images/SangerAlignment_ShinyApp_10.png" style="width: 693.6px; height: 323.4px;" /></a>
<p class="caption"><span class="caption-text">Figure 12. <em>SangerRead</em> page - Trimming Method 2 (M2): ‘Trimmomatics Sliding Window Trimming’ UI.</span><a class="headerlink" href="#id12" title="Permalink to this image">¶</a></p>
</div>
<p><a class="reference internal" href="#sangeralignment-shinyapp-11"><span class="std std-ref">Figure 13</span></a> shows the quality report before and after trimming. After clicking the “Apply Trimming Parameters” button, the values of these information boxes will be updated to the latest values.</p>
<div class="figure align-center" id="id13">
<span id="sangeralignment-shinyapp-11"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_11.png"><img alt="../_images/SangerAlignment_ShinyApp_11.png" src="../_images/SangerAlignment_ShinyApp_11.png" style="width: 654.3000000000001px; height: 389.7px;" /></a>
<p class="caption"><span class="caption-text">Figure 13. <em>SangerRead</em> page - read quality report before / after trimming.</span><a class="headerlink" href="#id13" title="Permalink to this image">¶</a></p>
</div>
<p>In <a class="reference internal" href="#sangeralignment-shinyapp-13"><span class="std std-ref">Figure 14</span></a>, the x-axis is the index of the base pairs; the y-axis is the Phred quality score. The green horizontal bar at the top of the plot is the raw read region and the orange horizontal bar represents the trimmed read region. Both <a class="reference internal" href="#sangeralignment-shinyapp-13"><span class="std std-ref">Figure 14</span></a> trimming plot and <a class="reference internal" href="#sangeralignment-shinyapp-14"><span class="std std-ref">Figure 15</span></a> chromatogram will be updated once users change the quality trimming parameters and click the “Apply Trimming Parameters&quot; button in <a class="reference internal" href="#sangeralignment-shinyapp-14"><span class="std std-ref">Figure 15</span></a>.</p>
<div class="figure align-center" id="id14">
<span id="sangeralignment-shinyapp-13"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_13.png"><img alt="../_images/SangerAlignment_ShinyApp_13.png" src="../_images/SangerAlignment_ShinyApp_13.png" style="width: 712.0px; height: 351.5px;" /></a>
<p class="caption"><span class="caption-text">Figure 14. <em>SangerRead</em> page - quality trimming plot.</span><a class="headerlink" href="#id14" title="Permalink to this image">¶</a></p>
</div>
<p>If we only see primary and secondary sequences in the table, we will loose some variations. Chromatogram is very helpful to check the peak resolution. <a class="reference internal" href="#sangeralignment-shinyapp-14"><span class="std std-ref">Figure 15</span></a> shows the panel of plotting chromatogram. Users can change four parameters: <code class="code docutils literal notranslate"><span class="pre">Base</span> <span class="pre">Number</span> <span class="pre">Per</span> <span class="pre">Row</span></code>, <code class="code docutils literal notranslate"><span class="pre">Height</span> <span class="pre">Per</span> <span class="pre">Row</span></code>, <code class="code docutils literal notranslate"><span class="pre">Signal</span> <span class="pre">Ratio</span> <span class="pre">Cutoff</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Show</span> <span class="pre">Trimmed</span> <span class="pre">Region</span></code>. Among them, <code class="code docutils literal notranslate"><span class="pre">Signal</span> <span class="pre">Ratio</span> <span class="pre">Cutoff</span></code> is the key parameter. If its value is default value 0.33, it indicates that the lower peak should be at least 1/3rd as high as the higher peak for it count as a secondary peak.</p>
<div class="figure align-center" id="id15">
<span id="sangeralignment-shinyapp-14"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_14.png"><img alt="../_images/SangerAlignment_ShinyApp_14.png" src="../_images/SangerAlignment_ShinyApp_14.png" style="width: 722.25px; height: 420.3px;" /></a>
<p class="caption"><span class="caption-text">Figure 15. <em>SangerRead</em> page - chromatogram panel.</span><a class="headerlink" href="#id15" title="Permalink to this image">¶</a></p>
</div>
<p>Here is an example of applying new chromatogram parameters. We click “Show Trimmed Region” to set its value from FALSE to TRUE. <a class="reference internal" href="#sangeralignment-shinyapp-15"><span class="std std-ref">Figure 16</span></a> shows the loading notification popup during base calling and chromatogram plotting.</p>
<div class="figure align-center" id="id16">
<span id="sangeralignment-shinyapp-15"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_15.png"><img alt="../_images/SangerAlignment_ShinyApp_15.png" src="../_images/SangerAlignment_ShinyApp_15.png" style="width: 725.4px; height: 423.90000000000003px;" /></a>
<p class="caption"><span class="caption-text">Figure 16. <em>SangerRead</em> page - loading notification popup during replotting chromatogram.</span><a class="headerlink" href="#id16" title="Permalink to this image">¶</a></p>
</div>
<p>After replotting the chromatogram, trimmed region is showed in red striped region. <a class="reference internal" href="#sangeralignment-shinyapp-16"><span class="std std-ref">Figure 17</span></a> shows part of the the chromatogram (1 bp ~ 240 bp). Moreover, chromatogram will be replotted when trimmed positions or chromatogram parameters are updated.</p>
<div class="figure align-center" id="id17">
<span id="sangeralignment-shinyapp-16"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_16.png"><img alt="../_images/SangerAlignment_ShinyApp_16.png" src="../_images/SangerAlignment_ShinyApp_16.png" style="width: 772.0px; height: 416.5px;" /></a>
<p class="caption"><span class="caption-text">Figure 17. <em>SangerRead</em> page - chromatogram with trimmed region showed.</span><a class="headerlink" href="#id17" title="Permalink to this image">¶</a></p>
</div>
<p>To let users browse the trimmed primary/secondary sequences without finding “Trimming Start Point” and “Trimming End Point” by themselves, we provide the final trimmed primary/secondary sequences that will be used for reads alignment in table format with quality scores in <a class="reference internal" href="#sangeralignment-shinyapp-17"><span class="std std-ref">Figure 18</span></a>. Frameshift amino acid sequences are also provided.</p>
<div class="figure align-center" id="id18">
<span id="sangeralignment-shinyapp-17"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_17.png"><img alt="../_images/SangerAlignment_ShinyApp_17.png" src="../_images/SangerAlignment_ShinyApp_17.png" style="width: 723.15px; height: 300.6px;" /></a>
<p class="caption"><span class="caption-text">Figure 18. <em>SangerRead</em> page - trimmed primary/secondary sequences and Phred quality score in table format.</span><a class="headerlink" href="#id18" title="Permalink to this image">¶</a></p>
</div>
<p>We have updated the trimming and chromatogram parameters for each read. Now, we need to click “Re-calculate contig” button to do alignment again. Last but not least, we can save all data into a new ‘SangerContig’ S4 instance by clicking “Save S4 instance button”. New S4 instance will be saved in <strong>Rda</strong> format. Users can run <code class="code docutils literal notranslate"><span class="pre">readRDS</span></code> function to load it into current R environment. <a class="reference internal" href="#sangeralignment-shinyapp-18"><span class="std std-ref">Figure 19</span></a> shows some hints in the save notification popup.</p>
<div class="figure align-center" id="id19">
<span id="sangeralignment-shinyapp-18"></span><a class="reference internal image-reference" href="../_images/SangerAlignment_ShinyApp_18.png"><img alt="../_images/SangerAlignment_ShinyApp_18.png" src="../_images/SangerAlignment_ShinyApp_18.png" style="width: 767.2px; height: 420.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Figure 19. <em>SangerRead</em> page - saving notification popup.</span><a class="headerlink" href="#id19" title="Permalink to this image">¶</a></p>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="writing-sangeralignment-fasta-files-ab1">
<h2>Writing <em>SangerAlignment</em> FASTA files <sub>(AB1)</sub><a class="headerlink" href="#writing-sangeralignment-fasta-files-ab1" title="Permalink to this headline">¶</a></h2>
<p>Users can write the <em>SangerAlignment</em> instance to <strong>FASTA</strong> files. There are four options for users to choose from in <code class="code docutils literal notranslate"><span class="pre">selection</span></code> parameter.</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">contigs_unalignment</span></code>: Writing contigs into a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">contigs_alignment</span></code>: Writing contigs alignment and contigs consensus read to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">all_reads</span></code>: Writing all reads to a single <strong>FASTA</strong> file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">all</span></code>: Writing contigs, contigs alignment, and all reads into three different files.</p></li>
</ul>
<p>Below is the one-line function that users need to run. This function mainly depends on <code class="code docutils literal notranslate"><span class="pre">writeXStringSet</span></code> function in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> R package. Users can set the compression level through <code class="code docutils literal notranslate"><span class="pre">writeFasta</span></code> function.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeFasta</span><span class="p">(</span><span class="n">sangerAlignment</span><span class="p">,</span>
           <span class="n">outputDir</span>         <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
           <span class="n">compress</span>          <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
           <span class="n">compression_level</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">,</span>
           <span class="n">selection</span>         <span class="o">=</span> <span class="s">&quot;all&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can download the output FASTA file of this example through the following three links:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/AB1/Sanger_contigs_unalignment.fa">contigs_unalignment FASTA file</a></p></li>
<li><p><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/AB1/Sanger_contigs_alignment.fa">contigs_alignment FASTA file</a></p></li>
<li><p><a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/AB1/Sanger_all_trimmed_reads.fa">all_reads FASTA file</a></p></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="generating-sangeralignment-report-ab1">
<h2>Generating <em>SangerAlignment</em> report <sub>(AB1)</sub><a class="headerlink" href="#generating-sangeralignment-report-ab1" title="Permalink to this headline">¶</a></h2>
<p>Last but not least, users can save <em>SangerAlignment</em> instance into a report after the analysis. The report will be generated in <strong>HTML</strong> by knitting <strong>Rmd</strong> files. There are two parameters, <code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code> and <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code>, for users to decide which level the <em>SangerAlignment</em> report will go. Moreover, after the reports are generated, users can easily navigate through reports in different levels within the <strong>HTML</strong> file.</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code>: Whether users want to generate the report of each <em>SangerContig</em> in <em>SangerAlignment</em>.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code>: If <code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code> is <code class="code docutils literal notranslate"><span class="pre">TRUE</span></code>, then users can set this value to decide whether they want to include <em>SangerRead</em> reports in each <em>SangerContig</em>.</p></li>
</ul>
<p>One thing to pay attention to is that if users have many reads, it will take quite a long time to write out all reports. If users only want to generate the contigs alignment, remember to set <code class="code docutils literal notranslate"><span class="pre">includeSangerContig</span></code> and <code class="code docutils literal notranslate"><span class="pre">includeSangerRead</span></code> to <code class="code docutils literal notranslate"><span class="pre">FALSE</span></code> in order to save time.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">generateReport</span><span class="p">(</span><span class="n">sangerAlignment</span><span class="p">,</span>
               <span class="n">outputDir</span>           <span class="o">=</span> <span class="nf">tempdir</span><span class="p">(),</span>
               <span class="n">includeSangerContig</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
               <span class="n">includeSangerRead</span>   <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>Users can access to '<em>Basic Information</em>', '<em>Contigs Consensus</em>', '<em>Contigs Alignment</em>' and '<em>Contigs Tree</em>' sections inside the generated <a class="reference external" href="https://howardchao.github.io/sangeranalyseR_report/SangerAlignment/AB1/SangerAlignment/SangerAlignment_Report.html">SangerContig html report of this example</a>. Furthermore, users can also navigate through html reports of all forward and reverse <em>SangerRead</em> in this <em>SangerContig</em> report.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="advance_user_SangerRead_fasta.html" class="btn btn-neutral float-right" title="Advanced User Guide - SangerRead (FASTA)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="advance_user_SangerContig_ab1.html" class="btn btn-neutral float-left" title="Advanced User Guide - SangerContig (AB1)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Kuan-Hao Chao

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>